econml/_cate_estimator.py

```diff
@@ -624,24 +629,49 @@ def marginal_effect(self, T, X=None):
             the corresponding singleton dimensions in the output will be collapsed
             (e.g. if both are vectors, then the output of this method will also be a vector)
         """
-        X, T = self._expand_treatments(X, T)
+        X, T = self._expand_treatments(X, T, transform=False)
         eff = self.const_marginal_effect(X)
-        return np.repeat(eff, shape(T)[0], axis=0) if X is None else eff
+
+        if X is None:
+            eff = np.repeat(eff, shape(T)[0], axis=0)
+
+        if hasattr(self, 'treatment_featurizer') and self.treatment_featurizer:
+            feat_T = self.transformer.transform(T)
+            jac_T = self.transformer.jac(T)
+
+            einsum_str = 'myf, mtf->myt'
+
+            if ndim(T) == 1:
+                einsum_str = einsum_str.replace('t', '')
+            if ndim(feat_T) == 1:
+                einsum_str = einsum_str.replace('f', '')
+            if (ndim(eff) == ndim(feat_T)):
+                einsum_str = einsum_str.replace('y', '')
+            return np.einsum(einsum_str, eff, jac_T)
+
+        else:
+            return eff
 
     def marginal_effect_interval(self, T, X=None, *, alpha=0.05):
-        X, T = self._expand_treatments(X, T)
-        effs = self.const_marginal_effect_interval(X=X, alpha=alpha)
-        if X is None:  # need to repeat by the number of rows of T to ensure the right shape
-            effs = tuple(np.repeat(eff, shape(T)[0], axis=0) for eff in effs)
-        return effs
+        if hasattr(self, 'treatment_featurizer') and self.treatment_featurizer:
+            return BaseCateEstimator._defer_to_inference(self.marginal_effect_interval)(self, T, X, alpha=alpha)
+        else:
+            X, T = self._expand_treatments(X, T)
+            effs = self.const_marginal_effect_interval(X=X, alpha=alpha)
+            if X is None:  # need to repeat by the number of rows of T to ensure the right shape
+                effs = tuple(np.repeat(eff, shape(T)[0], axis=0) for eff in effs)
+            return effs
     marginal_effect_interval.__doc__ = BaseCateEstimator.marginal_effect_interval.__doc__
 
     def marginal_effect_inference(self, T, X=None):
-        X, T = self._expand_treatments(X, T)
-        cme_inf = self.const_marginal_effect_inference(X=X)
-        if X is None:
-            cme_inf = cme_inf._expand_outputs(shape(T)[0])
-        return cme_inf
+        if hasattr(self, 'treatment_featurizer') and self.treatment_featurizer:
+            return BaseCateEstimator._defer_to_inference(self.marginal_effect_inference)(self, T, X)
```

Comment:
Can you elaborate on your suggestion here? I'm not sure what you mean by callable core.

Reply:
The thing that seems strange is that we're defining `marginal_effect_inference` in terms of `BaseCateEstimator._defer_to_inference(self.marginal_effect_inference)`, which initially seems like it might lead to an infinite loop (but it doesn't, because `_defer_to_inference` looks at the method being wrapped and then calls the method of the same name on the `_inference` attribute instead).

But it's also a little bit weird because `_defer_to_inference` is a decorator, and so directly calling it is a little bit unusual (there's some extra decorator logic created by `@wraps` that sets the method name, etc., that is all completely unnecessary here, since we just immediately call the wrapper with arguments, instead of assigning it to an attribute on the class like we normally would with a decorated instance).

So my thought was to split `_defer_to_inference` into something more like:

```
    def _use_inference_method(self, name, *args, **kwargs):
        if self._inference is not None:
            return getattr(self._inference, name)(*args, **kwargs)
        else:
            raise AttributeError("Can't call '%s' because 'inference' is None" % name)

    def _defer_to_inference(m):
        @wraps(m)
        def call(self, *args, **kwargs):
            self._use_inference_method(m.__name__, *args, **kwargs)
        return call
```

And then to use
```suggestion
            return self._use_inference_method('marginal_effect_inference', T, X)
```

which seems less likely to be confusing regarding whether it is actually recursive.